trademapper/js/trademapper.mapper.js

```diff
@@ -108,6 +104,116 @@ define(["d3", "topojson", "worldmap", "disputedareas", "countrycentre", "config"
 				.attr("fill", "url(#diagonalHatch)");
 	},
 
+	// load overlays from the overlays= variable in the query string
+	addOverlays: function () {
+		var params = (new URL(document.location)).searchParams;
+		var urls = params.get('overlays');
+
+		if (urls === null) {
+			return;
+		}
+
+		urls = urls.split(',');
+		console.log('FETCHING OVERLAYS FROM URLS:', urls);
+
+		urls.forEach(this.loadTopojsonFromURL.bind(this));
+	},
+
+	loadTopojsonFromURL: function(url) {
+		var self = this;
+
+		d3.json(url).then(
+			this.loadTopojson.bind(this),
+
+			function (error) {
+				console.log("unable to download", error);
+			}
+		);
+	},
+
+	loadTopojson: function (data) {
+		// lines need drawing with class overlay-line;
+		// polygons need drawing with class overlay-area, but also need to
+		// add lines denoting their boundaries;
+		// points can be treated the same way as polygons and don't need boundaries
+		// to be derived
+		var polygons = {
+			type: "GeometryCollection",
+			geometries: []
+		};
+
+		var lines = {
+			type: "GeometryCollection",
+			geometries: []
+		};
+
+		var points = {
+			type: "GeometryCollection",
+			geometries: []
+		};
+
+		// sort the geometries so they can be drawn appropriately
+		var geometries, i, geometry;
+		for (var propertyName in data.objects) {
+			geometries = data.objects[propertyName].geometries;
+
+			for (i = 0; i < geometries.length; i++) {
+				var geometry = geometries[i];
+
+				if (geometry.type === "Polygon" || geometry.type === "MultiPolygon") {
+					polygons.geometries.push(geometry);
+				} else if (geometry.type === "LineString" || geometry.type === "MultiLineString") {
+					lines.geometries.push(geometry);
+				} else if (geometry.type === "Point" || geometry.type === "MultiPoint") {
+					points.geometries.push(geometry);
+				}
+			}
+		}
+
+		if (polygons.geometries.length > 0) {
+			var polygonFeatures = topojson.feature(data, polygons).features;
+
+			this.mapg.selectAll(".overlay-polygon")
```

Comment:
This doesn't fully make sense to me, but that's probably because I don't know d3. Where is the element with class "overlay-polygon" that you select here created? I've greped the source code but couldn't find it.

I see you add class "overlay-polygon" below, but starting from an empty set of elements, what would this add it to? And why add it again if you've just selected it?

So I assume this works and it's just I don't understand d3, but if you explain it then at least I'll learn :)

Reply:
It's a bit odd, granted. The overlay-polygon elements are actually being created here. Using selectAll() with a selector that has no matching elements creates an empty group which can then be populated. I don't understand the semantics in depth, but [this blog post](http://knowledgestockpile.blogspot.com/2012/01/understanding-selectall-data-enter.html) goes into the details.